access/jira/app.go

```diff
@@ -297,82 +375,272 @@ func (a *App) onJIRAWebhook(ctx context.Context, webhook Webhook) error {
+		if issueID == "" {
+			return PluginData{}, false
+		}
+
+		count := len(reqReviews)
+		if oldCount = existing.ReviewsCount; oldCount >= count {
+			return PluginData{}, false
 		}
+		pluginData := *existing
+		pluginData.ReviewsCount = count
+		return pluginData, true
+	})
+	if err != nil {
 		return trace.Wrap(err)
 	}
+	if !ok {
+		if issueID == "" {
+			logger.Get(ctx).Debug("Failed to add the comment: plugin data is blank")
+		}
+		return nil
+	}
+	ctx, _ = logger.WithField(ctx, "jira_issue_id", issueID)
 
-	reqData, jiraData := pluginData.RequestData, pluginData.JiraData
-	if jiraData.ID == "" {
-		log.Warn("Plugin data is either missing or expired")
+	slice := reqReviews[oldCount:]
+	if len(slice) == 0 {
 		return nil
 	}
 
-	if err := a.bot.ExpireIssue(ctx, reqID, reqData, jiraData); err != nil {
+	errors := make([]error, 0, len(slice))
+	for _, review := range slice {
+		if err := a.jira.AddIssueReviewComment(ctx, issueID, review); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// resolveIssue transitions the issue to some final state.
+func (a *App) resolveIssue(ctx context.Context, reqID string, resolution Resolution) error {
+	var issueID string
+
+	// Save request resolution info in plugin data.
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		// If plugin data is empty or missing issueID, we cannot do anything.
+		if existing == nil {
+			issueID = ""
+			return PluginData{}, false
+		}
+
+		issueID = existing.IssueID
+		// If plugin data has blank issue identification info, we cannot do anything.
+		if issueID == "" {
+			return PluginData{}, false
+		}
+
+		// If resolution field is not empty then we already resolved the issue before. In this case we just quit.
+		if existing.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
+		}
+
+		// Mark issue as resolved.
+		pluginData := *existing
+		pluginData.Resolution = resolution
+		return pluginData, true
+	})
+	if err != nil {
 		return trace.Wrap(err)
 	}
+	if !ok {
+		if issueID == "" {
+			logger.Get(ctx).Debug("Failed to resolve the issue: plugin data is blank")
+		}
 
-	log.Info("Successfully marked request as expired")
+		// Either plugin data is missing or issue is already resolved by us, just quit.
+		return nil
+	}
+
+	ctx, log := logger.WithField(ctx, "jira_issue_id", issueID)
+	if err := a.jira.ResolveIssue(ctx, issueID, resolution); err != nil {
+		return trace.Wrap(err)
+	}
+	log.Info("Successfully resolved the issue")
 
 	return nil
 }
 
-func (a *App) getPluginData(ctx context.Context, reqID string) (PluginData, error) {
-	dataMap, err := a.accessClient.GetPluginData(ctx, reqID)
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
```

Comment:
Might be worth mentioning in the docstring that `fn` may be invoked many times and shouldn't do anything with a heap of side-effects. Also might be worth mentioning _why_ it needs to be a compare-and-set operation as well. 

(I don't know the plugin API well, so I assume its because there is a chance someone could sweep in and alter the server-side plugin config while this update is in transit?)

Reply:
Design question - would it make sense to keep trying to modify the config until it succeeds (or fails with a non-`CompareFailed` error) or the supplied `ctx`expires?

That way you can have one, top-level timeout controlling how long to try and start the plugin, rather than arbitrary retries scattered throughout the code.

Just a suggestion.

Reply:
Also, it would be nice to have what the update function documented (e.g., what does returning `false` imply for the behaviout of `modifyPluginData()`.

Reply:
> I don't know the plugin API well, so I assume its because there is a chance someone could sweep in and alter the server-side plugin config while this update is in transit?

Yes cause there's some gap between `GetAccessRequests(...)` and `UpdatePluginData(...)` calls.

Reply:
> Design question - would it make sense to keep trying to modify the config until it succeeds (or fails with a non-CompareFailed error) or the supplied ctx expires?

Oh, after some hesitation, I rewrote `modifyPluginData` to use exponential backoff!